File: dynamo/Python/src/NaaN/dynamo/Geometry/DiamondPoints.py
import math
from ... import utils as nutils
from .pythonPoint import pythonPoint


class DiamondPoints(object):
    def __init__(self, xnum, ynum, xdis, ydis):
        self.xnum = xnum
        self.ynum = ynum
        self.xdis = xdis
        self.ydis = ydis
        self.twoLayerPoints = self.__genDiamondLayer_T1()
        self.AllPanels = self.__DIAGET()
        self.DiamondOuter = self.AllPanels[0]
        self.DiamondInner = self.AllPanels[1]
        self.TriangleA = self.AllPanels[2]
        self.TriangleB = self.AllPanels[3]
        self.TriangleC = self.AllPanels[4]
        self.TriangleD = self.AllPanels[5]

    def DiamondInfo(self):
        return [self.xnum, self.ynum, self.xdis, self.ydis]

    def __DIA_getPoint(self, x, y):
        return pythonPoint(x, y, 0)

    def __DIA_genlayer(self, xnum, ynum, xdis, ydis, offset):
        ret = []
        for i in range(0, xnum * xdis, xdis):
            xlayer = []
            for j in range(0, ynum * ydis, ydis):
                t = self.__DIA_getPoint(i + offset.x, j + offset.y)
                xlayer.append(t)
            ret.append(xlayer)
        return ret

    def __genDiamondLayer_T1(self):
        halfwidth = self.xdis / 2
        halfheight = self.ydis / 2
        firstLayer = self.__DIA_genlayer(self.xnum, self.ynum, self.xdis,
                                         self.ydis, pythonPoint(0, 0))
        secondLayer = self.__DIA_genlayer(self.xnum - 1, self.ynum - 1,
                                          self.xdis, self.ydis,
                                          pythonPoint(halfwidth,
                                                      halfheight))
        return [firstLayer, secondLayer]

    def __DIA_getTriangles(self, TP):
        fl = TP[0]
        sl = TP[1]
        triangleA = []
        triangleB = []
        for i in range(0, self.xnum - 1):
            triangleA.append([fl[i][0], fl[i+1][0], sl[i][0]])
            triangleB.append([fl[i][self.ynum-1], fl[i+1][self.ynum-1],
                              sl[i][self.ynum-2]])
        triangleC = []
        triangleD = []
        for i in range(0, self.ynum - 1):
            triangleC.append([fl[self.xnum-1][i], fl[self.xnum-1][i+1],
                              sl[self.xnum-2][i]])
            triangleD.append([fl[0][i], fl[0][i+1], sl[0][i]])
        return [triangleA, triangleB, triangleC, triangleD]

    def __DIA_getDiamond(self, TP):
        fl = TP[0]
        sl = TP[1]
        diamondA = []
        diamondB = []
        for i in range(0, self.ynum - 1):
            for j in range(0, self.xnum - 1):
                diamondA.append([fl[j][i], sl[j-1][i], fl[j][i+1], sl[j][i]])
        for i in range(0, self.xnum - 1):
            for j in range(1, self.ynum - 1):
                diamondB.append([fl[i][j], sl[i][j-1], fl[i+1][j], sl[i][j]])
        return [diamondA, diamondB]

    def __DIAGET(self):
        return (self.__DIA_getDiamond(self.twoLayerPoints) +
                self.__DIA_getTriangles(self.twoLayerPoints))

    def TransformPoints(self, RX, RY, RateX=0.3, RateY=0.3):
        if(RateX < 0 or RateX > 1 or RateY < 0 or RateY > 1):
            raise Exception("RateX and RateY must in interval [0, 1]")
        nutils.multiLevelMap(
            lambda x: x.TransformX(RX, RY,
                                   RateX * self.xdis * self.xnum,
                                   RateY * self.ydis * self.ynum),
            self.twoLayerPoints)
        return self

    def DefaultTransformPoints(self):
        return self.TransformPoints(self.xdis * 4, self.ydis * 4)

    def __repr__(self):
        return (("{0}.DiamondPoints(xnum = {1}, " +
                 "ynum = {2}, xdis = {3}, ydis = {4})").
                format(
                __package__,
                self.xnum.__repr__(),
                self.ynum.__repr__(),
                self.xdis.__repr__(),
                self.ydis.__repr__()))

    def __str__(self):
        return str(nutils.multiLevelMap(
            lambda x: x.__str__(),
            self.DefaultTransformPoints().AllPanels))

File: dynamo/csharp/src/Geometry/DiamondPoints.cs
using System.Collections.Generic;
using System.Linq;
using NaaN;

namespace NaaN.dynamo.Geometry
{
    public class DiamondPoints
    {
        private int xnum, ynum;
        private double xdis, ydis;
        private List<List<List<CSPoint>>> twoLayerPoints;
        private List<List<List<CSPoint>>> allPanels;

        public List<List<List<CSPoint>>> AllPanels { get { return this.allPanels; } }
        public List<List<CSPoint>> DiamondY { get { return this.allPanels[0]; } }
        public List<List<CSPoint>> DiamondX { get { return this.allPanels[1]; } }
        public List<List<CSPoint>> TriangleA { get { return this.allPanels[2]; } }
        public List<List<CSPoint>> TriangleB { get { return this.allPanels[3]; } }
        public List<List<CSPoint>> TriangleC { get { return this.allPanels[4]; } }
        public List<List<CSPoint>> TriangleD { get { return this.allPanels[5]; } }

        public DiamondPoints(int xnum, int ynum, double xdis, double ydis)
        {
            this.xnum = xnum;
            this.ynum = ynum;
            this.xdis = xdis;
            this.ydis = ydis;
            this.twoLayerPoints = __genDiamondLayer_T1();
            this.allPanels = __DIAGET();
        }
        private DiamondPoints() { }

        public System.Collections.ArrayList DiamondInfo
        {
            get
            {
                return new System.Collections.ArrayList() { this.xnum, this.ynum, this.xdis, this.ydis };
            }
        }

        private List<List<CSPoint>> __DIA_genLayer(
            int xnum, int ynum,
            double xdis, double ydis, CSPoint offset)
        {
            List<List<CSPoint>> ret = new List<List<CSPoint>>();
            foreach (int i in Enumerable.Range(0, xnum))
            {
                List<CSPoint> xlayer = new List<CSPoint>();
                foreach (int j in Enumerable.Range(0, ynum))
                {
                    xlayer.Add(new CSPoint(i * xdis, j * ydis) + offset);
                }
                ret.Add(xlayer);
            }
            return ret;
        }

        private List<List<List<CSPoint>>> __genDiamondLayer_T1()
        {
            double halfWidth = this.xdis / 2;
            double halfHeight = this.ydis / 2;
            List<List<CSPoint>> firstLayer = this.__DIA_genLayer(this.xnum, this.ynum, this.xdis, this.ydis, CSPoint.CSPointOrigin);
            List<List<CSPoint>> secondLayer = this.__DIA_genLayer(this.xnum - 1, this.ynum - 1,
                                                                  this.xdis, this.ydis, new CSPoint(halfWidth, halfHeight));
            return new List<List<List<CSPoint>>>() { firstLayer, secondLayer };
        }

        private List<List<List<CSPoint>>> __DIA_getTriangles(List<List<List<CSPoint>>> TP)
        {
            List<List<CSPoint>> fl = TP[0];
            List<List<CSPoint>> sl = TP[1];
            List<List<CSPoint>> triangleA = new List<List<CSPoint>>();
            List<List<CSPoint>> triangleB = new List<List<CSPoint>>();
            List<List<CSPoint>> triangleC = new List<List<CSPoint>>();
            List<List<CSPoint>> triangleD = new List<List<CSPoint>>();
            foreach (int i in Enumerable.Range(0, this.xnum - 1))
            {
                triangleA.Add(new List<CSPoint>() { fl[i][0], fl[i + 1][0], sl[i][0] });
                triangleB.Add(new
                    List<CSPoint>() { fl[i][this.ynum - 1], fl[i + 1][this.ynum - 1], sl[i][this.ynum - 2] });
            }

            foreach (int i in Enumerable.Range(0, this.ynum - 1))
            {
                triangleC.Add(new
                    List<CSPoint>() { fl[this.xnum - 1][i], fl[this.xnum - 1][i + 1], sl[this.xnum - 2][i] });
                triangleD.Add(new List<CSPoint>() { fl[0][i], fl[0][i + 1], sl[0][i] });
            }
            return new List<List<List<CSPoint>>>() { triangleA, triangleB, triangleC, triangleD };
        }

        private List<List<List<CSPoint>>> __DIA_getDiamond(List<List<List<CSPoint>>> TP)
        {
            List<List<CSPoint>> fl = TP[0];
            List<List<CSPoint>> sl = TP[1];
            List<List<CSPoint>> diamondA = new List<List<CSPoint>>();
            List<List<CSPoint>> diamondB = new List<List<CSPoint>>();
            foreach (int i in Enumerable.Range(0, this.ynum - 1))
            {
                foreach (int j in Enumerable.Range(1, this.xnum - 2))
                {
                    diamondA.Add(new List<CSPoint>() { fl[j][i], sl[j - 1][i], fl[j][i + 1], sl[j][i] });
                }
            }
            foreach (int i in Enumerable.Range(0, this.xnum - 1))
            {
                foreach (int j in Enumerable.Range(1, this.ynum - 2))
                {
                    diamondB.Add(new List<CSPoint>() { fl[i][j], sl[i][j - 1], fl[i + 1][j], sl[i][j] });
                }
            }
            return new List<List<List<CSPoint>>>() { diamondA, diamondB };
        }

        private List<List<List<CSPoint>>> __DIAGET()
        {
            List<List<List<CSPoint>>> D = this.__DIA_getDiamond(this.twoLayerPoints);
            List<List<List<CSPoint>>> T = this.__DIA_getTriangles(this.twoLayerPoints);
            return new List<List<List<CSPoint>>>(Enumerable.Concat(D, T));
        }

        public void TransformPoints(double RX, double RY, double RateX = 0.3, double RateY = 0.3)
        {
            if (RateY < 0 || RateX < 0 || RateY > 1 || RateX > 1)
            {
                throw new System.ArgumentException("RateX and RateY must in interval [0, 1]");
            }
            utils.multiLevelMap(
                x =>
                {
                    CSPoint cxpt = x as CSPoint;
                    cxpt.TransformX(RX, RY,
                    this.xnum * this.xdis * RateX,
                    this.ynum * this.ydis * RateY);
                    return 0;
                }, this.twoLayerPoints);
            return;
        }

        public override bool Equals(object obj) { return false; }
        public bool Equals(DiamondPoints obj) { return false; }
        public override int GetHashCode()
        {
            return (ynum % (xnum % 0x20)) * (int)(xdis / ydis);
        }

        public override string ToString()
        {
            return allPanels.ToString();
        }
    }
}